cli/src/description_util.rs

```diff
@@ -31,12 +43,138 @@ JJ: Lines starting with "JJ: " (like this one) will be removed.
+    pub missing: Vec<String>,
+    /// Commit IDs that were found multiple times while parsing the edited
+    /// messages.
+    pub duplicates: Vec<String>,
+    /// Commit IDs that were found while parsing the edited messages, but which
+    /// were not originally being edited.
+    pub unexpected: Vec<String>,
+}
+
+/// Edits the descriptions of the given commits in a single editor session.
+pub fn edit_multiple_descriptions(
+    tx: &mut WorkspaceCommandTransaction,
+    repo: &ReadonlyRepo,
+    commits: &[&Commit],
+    settings: &UserSettings,
+) -> Result<EditMultipleDescriptionsResult, CommandError> {
+    let mut commits_map = HashMap::new();
+    let mut output_chunks = Vec::new();
+
+    for &commit in commits.iter() {
+        let commit_hash = short_commit_hash(commit.id());
+        if commits.len() > 1 {
+            output_chunks.push(format!("JJ: describe {} -------\n", commit_hash.clone()));
+        }
+        commits_map.insert(commit_hash, commit.id());
+        let mut commit_builder = tx.mut_repo().rewrite_commit(settings, commit).detach();
+        if commit_builder.description().is_empty() {
+            commit_builder.set_description(settings.default_description());
+        }
+        let temp_commit = commit_builder.write_hidden()?;
+        let template = description_template(tx, "", &temp_commit)?;
+        output_chunks.push(template);
+        output_chunks.push("\n".to_owned());
+    }
+    output_chunks
+        .push("JJ: Lines starting with \"JJ: \" (like this one) will be removed.\n".to_owned());
+    let bulk_message = output_chunks.join("");
+
+    let bulk_message = edit_temp_file(
+        "description",
+        ".jjdescription",
+        repo.repo_path(),
+        &bulk_message,
+        settings,
+    )?;
+
+    if commits.len() == 1 {
+        return Ok(EditMultipleDescriptionsResult {
+            descriptions: HashMap::from([(
+                commits[0].id().clone(),
+                cleanup_description(&bulk_message),
+            )]),
+            missing: vec![],
+            duplicates: vec![],
+            unexpected: vec![],
+        });
+    }
+
+    Ok(parse_bulk_edit_message(&bulk_message, &commits_map))
+}
+
+/// Parse the bulk message of edited commit descriptions.
+fn parse_bulk_edit_message(
+    message: &str,
+    commit_ids_map: &HashMap<String, &CommitId>,
+) -> EditMultipleDescriptionsResult {
+    let mut descriptions = HashMap::new();
+    let mut duplicates = Vec::new();
+    let mut unexpected = Vec::new();
+
+    let messages = message.lines().fold(vec![], |mut accum, line| {
+        if let Some(commit_id_prefix) = line.strip_prefix("JJ: describe ") {
+            let commit_id_prefix = commit_id_prefix
+                .strip_suffix(" -------")
+                .unwrap_or(commit_id_prefix);
+            accum.push((commit_id_prefix, vec![]));
+        } else if let Some((_, lines)) = accum.last_mut() {
+            lines.push(line);
+        };
```

Comment:
nit: `} else {` error?

Reply:
I'm not sure if that's strictly necessary, because I'm not sure how to nicely error out from the `fold` call. Also it should probably be handled by the code for missing entries.

Reply:
Hmm, I think non-empty content line without `JJ: describe` header should be an error because ignoring error can lead to data loss. Perhaps, it could be mapped to an entry of `commit_id_prefix == ""`?

It might be even easier if `parse_bulk_edit_message()` can return error. We don't have to use `fold()`.

Reply:
Updated to do this.